src/kudu/tserver/consensus_service.cc

```diff
@@ -214,6 +214,44 @@ bool GetConsensusOrRespond(TSTabletManager* tablet_manager,
   return true;
 }
 
+template <class ReqType, class RespType>
+bool CheckRaftRpcTokenOrRespond(const std::string &method_name,
+                                const ReqType *req, RespType resp,
+                                rpc::RpcContext *context,
+                                const consensus::RaftConsensus &consensus) {
+  const auto &ownToken = consensus.GetRaftRpcToken();
+  if (!ownToken && !req->has_raft_rpc_token()) {
+    // Empty on both, nothing to enforce
+    return true;
+  }
+
+  if (ownToken && req->has_raft_rpc_token() &&
+      *ownToken == req->raft_rpc_token()) {
+    // Tokens match
+    return true;
+  }
+
+  auto error_message = Substitute(
+      "Raft RPC token mismatch. Receiver token: $0. Request token: $1",
+      ownToken ? *ownToken : "<null>",
+      req->has_raft_rpc_token() ? req->raft_rpc_token() : "<null>");
+
+  if (!consensus.ShouldEnforceRaftRpcToken()) {
+    // Mismatch but don't enforce
+    KLOG_EVERY_N(WARNING, 5000)
```

Comment:
would time based be less chatty?

Reply:
Probably similar because our heartbeats are regular enough, but I like the idea because it'll be more predictable.